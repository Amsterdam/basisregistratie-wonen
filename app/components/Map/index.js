import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import amaps from 'amsterdam-amaps/dist/amaps';
import 'amsterdam-amaps/dist/nlmaps/dist/assets/css/nlmaps.css';
import 'leaflet/dist/leaflet.css';

import { rdToWgs84 } from 'shared/services/crs-converter/crs-converter';

export const MapWrapper = styled.div`
  height: 200px;
  width: 100%;
  position: relative;
  z-index: 0;
`;

export const target = 'mapDiv';

const Map = ({ center, coordinates, marker, search, zoom }) => {
  useEffect(() => {
    let points = center;

    if (coordinates) {
      points = rdToWgs84(coordinates);
    }

    if (points) {
      amaps.createMap({
        center: points,
        marker,
        search,
        target,
        zoom,
      });
    }

    return () => {
      // clear mapDiv contents; amaps creates a new placeholder with every component re-render
      const mapDiv = document.getElementById(target);
      while (mapDiv.firstChild) {
        mapDiv.removeChild(mapDiv.firstChild);
      }
    };
  }, [coordinates, center]);

  return (
    <section>
      <MapWrapper className="cf">
        <div id={target} style={{ height: '100%' }} />
      </MapWrapper>
    </section>
  );
};

Map.defaultProps = {
  center: null,
  coordinates: null,
  marker: false,
  search: true,
  zoom: 13,
};

Map.propTypes = {
  /**
   * Coordinates in RD-format (https://nl.wikipedia.org/wiki/Rijksdriehoeksco%C3%B6rdinaten)
   * @see {@link https://nl.wikipedia.org/wiki/Rijksdriehoeksco%C3%B6rdinaten}
   */
  coordinates: PropTypes.shape({
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
  }),
  /** Latitude/longitude coordinates. When used in conjunction with `coordinates`, `coordinates` will take precedence. */
  center: PropTypes.shape({
    latitude: PropTypes.number.isRequired,
    longitude: PropTypes.number.isRequired,
  }),
  /** When false, no search input field will be shown */
  search: PropTypes.bool,
  /** Zoom level that is applied to the visible area of the map. Maximum zoom level: 22, minimum zoom level: 11. */
  zoom: PropTypes.number,
  /** When true, a marker will be shown at the position that is indicated by either `center` or `coordinates` */
  marker: PropTypes.bool,
};

export default Map;
