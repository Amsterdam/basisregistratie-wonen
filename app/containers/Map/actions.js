import { SEARCH_SELECT } from './constants';

export const searchSelect = payload => ({
  type: SEARCH_SELECT,
  payload,
});
