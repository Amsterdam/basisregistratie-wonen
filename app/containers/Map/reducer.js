import produce from 'immer';
import { SEARCH_SELECT } from './constants';

// The initial state of the App
export const initialState = {
  loading: false,
  error: false,
};

/* eslint-disable default-case, no-param-reassign */
export default (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case SEARCH_SELECT:
        draft.latlng = action.payload.latlng;
        draft.location = action.payload.location;
        draft.resultObject = action.payload.resultObject;
        break;
    }
  });
