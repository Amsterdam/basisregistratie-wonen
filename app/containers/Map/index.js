import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { makeSelectCoordinates } from 'containers/Verblijfsobject/selectors';
import Map from 'components/Map';

const mapStateToProps = createStructuredSelector({
  coordinates: makeSelectCoordinates,
});

const withConnect = connect(mapStateToProps);

export default compose(withConnect)(Map);
