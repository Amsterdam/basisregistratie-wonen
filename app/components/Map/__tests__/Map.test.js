import React from 'react';
import { render, cleanup } from 'react-testing-library';
import amaps from 'amsterdam-amaps/dist/amaps';
import { rdToWgs84 } from 'shared/services/crs-converter/crs-converter';
import 'jest-styled-components';
import Map from '..';

jest.mock('amsterdam-amaps/dist/amaps', () => {
  const actual = require.requireActual('amsterdam-amaps/dist/amaps');

  return {
    ...actual,
    createMap: jest.fn(),
  };
});

jest.mock('shared/services/crs-converter/crs-converter', () => {
  const actual = require.requireActual('shared/services/crs-converter/crs-converter');

  return {
    ...actual,
    rdToWgs84: jest.fn(() => ({ latitude: 0.0, longitude: 0.0 })),
  };
});

describe('Map', () => {
  const coordinates = {
    x: 123213,
    y: 488794,
  };
  const center = {
    latitude: 52.3547322,
    longitude: 4.8285841,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(cleanup);

  it('matches the snapshot', () => {
    const { container } = render(<Map center={center} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a map', () => {
    render(<Map coordinates={coordinates} marker={false} search={false} zoom={20} />);

    expect(amaps.createMap).toHaveBeenCalledWith(
      expect.objectContaining({
        marker: false,
        search: false,
        zoom: 20,
      }),
    );
  });

  it('converts points to lat/lng', () => {
    const { rerender } = render(<Map center={center} />);

    expect(rdToWgs84).not.toHaveBeenCalled();

    rerender(<Map coordinates={coordinates} />);

    expect(rdToWgs84).toHaveBeenCalledWith(coordinates);
  });
});
