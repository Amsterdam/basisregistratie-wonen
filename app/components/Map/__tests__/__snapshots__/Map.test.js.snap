// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map matches the snapshot 1`] = `
<section>
  <div
    class="cf Map__MapWrapper-sc-1v5i1wm-0 dwptan"
  >
    <div
      id="mapDiv"
      style="height: 100%;"
    />
  </div>
</section>
`;
